### PR TITLE
drivers/i2c: add i2c driver on it8xxx2 platform

### PR DESCRIPTION
**Pull request by GTLin08**
This commit is about the it8xxx2 i2c master driver which
includes six SMBus channels. The enhanced channel i2c3,
i2c4, i2c5 are controller which are designed to support
the I2C protocol.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

### CODE REVIEW

File: CODEOWNERS [hunk 1]
```diff
@@ -206,8 +206,10 @@
 /drivers/i2s/i2s_ll_stm32*                @avisconti
 /drivers/i2c/i2c_common.c                 @sjg20
 /drivers/i2c/i2c_emul.c                   @sjg20
+/drivers/i2c/i2c_ite_it8xxx2.c            @ite
```
Comment (pabigot) [inline CODEOWNERS:209]: <at>ite is not a github handle.  Who is the codeowner?

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 2]
```diff
@@ -0,0 +1,873 @@
...
+#define I2C_STANDARD_PORT_COUNT 3
+/* Default PLL frequency. */
+#define PLL_CLOCK 48000000
+
+struct i2c_it8xxx2_config {
+	uint8_t *base;
+	void (*irq_config_func)(void);
+	uint32_t bitrate;
+	uint32_t i2c_irq_base;
+	int port;
```
Comment (pabigot) [inline drivers/i2c/i2c_ite_it8xxx2.c:31]: This seems to be an ordinal identifying the port.  `uint8_t` would be more appropriate; it shouldn't be signed.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 3]
```diff
@@ -0,0 +1,873 @@
...
+	I2C_CH_WAIT_READ,
+	I2C_CH_WAIT_NEXT_XFER,
+};
+
+struct i2c_it8xxx2_data {
+	struct k_sem device_sync_sem;
+	uint16_t addr_16bit;
+	struct i2c_msg *msgs;
+	int err;			 /* Error code, if any */
+	int widx;			 /* Index into output data */
```
Comment (pabigot) [inline drivers/i2c/i2c_ite_it8xxx2.c:46]: If these indexes are non-negative, use an unsigned type.  If there's hardware constraints on the maximum offset use a smaller type; otherwise use `size_t`.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 4]
```diff
@@ -0,0 +1,873 @@
...
+	int ridx;			 /* Index into input data */
+	enum i2c_ch_status i2ccs;
+	uint8_t freq;        /* Frequency setting */
```
Comment (pabigot) [inline drivers/i2c/i2c_ite_it8xxx2.c:49]: Order the fields so there's no alignment-required internal packing. I.e. put structs and pointers first, then 32-bit values, then 16-bit values, then 8-bit values.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 5]
```diff
@@ -0,0 +1,873 @@
...
+	{ &GPDMRA, &GPDMRA,	0x10, 0x20},
+};
+
+static int i2c_get_line_levels(const struct device *dev)
+{
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+	int pin_sts = 0;
+
+	if (config->port < I2C_STANDARD_PORT_COUNT)
```
Comment (pabigot) [inline drivers/i2c/i2c_ite_it8xxx2.c:140]: Braces around all single-line code blocks.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 6]
```diff
@@ -0,0 +1,873 @@
...
+}
+
+/* Set clock frequency for i2c port D, E , or F */
+static void i2c_enhanced_port_set_frequency(const struct device *dev,
+				int freq_khz)
+{
+	struct i2c_it8xxx2_data *data = DEV_DATA(dev);
+	const struct i2c_it8xxx2_config *config = DEV_CFG(dev);
+	uint8_t *base = config->base;
+	int clk_div, psr;
```
Comment (pabigot) [inline drivers/i2c/i2c_ite_it8xxx2.c:229]: Using signed values is probably wrong here, and could introduce undefined behavior with overflows.  These should be unsigned, and the integer constants in the calculations below have `U` suffices (e.g. `2U * 1000U ...) - 2U;`)

**Review by pabigot — DISMISSED** [review]
First pass high-level comments.  The main concern is with use of `int` where values cannot be negative, and the required range is probably only 8- or 16-bit.  Zephyr has had multiple problems resulting from using signed types to hold non-negative values; I'd like to avoid introducing more.

**Comment by jettr** [comment]
@sjg20 @keith-zephyr @jackrosenthal please review

File: drivers/i2c/Kconfig.it8xxx2 [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2020 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+config I2C_ITE_IT8XXX2
+	bool "ITE IT8XXX2 I2C driver"
+	help
+	  Enable I2C support on it8xxx2_evb
```
Comment (sjg20) [inline drivers/i2c/Kconfig.it8xxx2:7]: Please expand this to explain

- speeds supported
- repeated start support
- anything else useful?

File: dts/bindings/i2c/ite,it8xxx2-i2c.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    reg:
+      required: true
+
+    label:
+      required: true
+
+    interrupts:
+      required: true
+
+    port_num:
```
Comment (sjg20) [inline dts/bindings/i2c/ite,it8xxx2-i2c.yaml:20]: missing description?

File: dts/riscv/it8xxx2.dtsi [hunk 9]
```diff
@@ -155,5 +156,71 @@
...
+		i2c0: i2c@f01c40 {
+			compatible = "ite,it8xxx2-i2c";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x00f01c40 0x0040>;
+			interrupts = <9 IRQ_TYPE_EDGE_FALLING>;
+			interrupt-parent = <&intc>;
+			status = "disabled";
+			label = "I2C_0";
+			port_num = <0>;
```
Comment (sjg20) [inline dts/riscv/it8xxx2.dtsi:168]: port-num

File: dts/bindings/i2c/ite,it8xxx2-i2c.yaml (hunk 8 above)
Comment (sjg20) [inline dts/bindings/i2c/ite,it8xxx2-i2c.yaml:20]: port-num

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 10]
```diff
@@ -833,7 +833,7 @@
 
 #define GCR2			ECREG(EC_REG_BASE_ADDR + 0x16F1)
 #define CK32OE			BIT(6)
-#define SMB4E			BIT(5)
+#define SMB3E			BIT(5)
```
Comment (sjg20) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:836]: What is happening here? If this is a bug fix, can you please put it in a separate commit (same PR)?

**Review by sjg20 — CHANGES_REQUESTED** [review]

File: CODEOWNERS (hunk 1 above)
Comment (jackrosenthal) [reply CODEOWNERS:209]: +1

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 10 above)
Comment (GTLin08) [reply soc/riscv/riscv-ite/common/chip_chipregs.h:836]: Done

File: dts/bindings/i2c/ite,it8xxx2-i2c.yaml (hunk 8 above)
Comment (GTLin08) [reply dts/bindings/i2c/ite,it8xxx2-i2c.yaml]: Done

Comment (GTLin08) [reply dts/bindings/i2c/ite,it8xxx2-i2c.yaml]: Done

File: dts/riscv/it8xxx2.dtsi (hunk 9 above)
Comment (GTLin08) [reply dts/riscv/it8xxx2.dtsi]: Done

File: drivers/i2c/Kconfig.it8xxx2 (hunk 7 above)
Comment (GTLin08) [reply drivers/i2c/Kconfig.it8xxx2]: Done

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 2 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 3 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 4 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 6 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 5 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done

**Review by pabigot — COMMENTED** [review]
Current version looks OK, no objections.  Without access to hardware/documentation I'm not comfortable approving this.

**Comment by GTLin08** [comment]
@pabigot 
In the index.rst, there are some description about it8xxx2 board,
or the following path has related instructions.
http://www.ite.com.tw/en/product/view?mid=149

**Review by sjg20 — APPROVED** [review]

**Review by jackrosenthal — APPROVED** [review]
From a high level perspective, it's looking good. I wasn't able to go too much into the specifics of the large C file as it was quite large.

**Comment by GTLin08** [comment]
I'm rebased code to master and fix merger conflict.
